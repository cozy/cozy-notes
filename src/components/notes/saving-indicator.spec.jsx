import React from 'react'
import { render } from '@testing-library/react'
import useBreakpoints from 'cozy-ui/transpiled/react/providers/Breakpoints'
import SavingIndicator from './saving-indicator'
import { AppLike } from '../../../test/AppLike'

jest.mock('cozy-ui/transpiled/react/providers/Breakpoints')

const sec = 1000
const min = 60 * sec
const hour = 60 * min
const day = 24 * hour

const getDirtySince = jest.fn()
const isDirty = () => !!getDirtySince()
const getLastSaveOrSync = jest.fn()
const on = jest.fn()

const collabProvider = {
  isDirty,
  getDirtySince,
  getLastSaveOrSync,
  on,
  off: jest.fn()
}

global.document.createRange = jest.fn()

function itMatchSnapshot(collabProvider) {
  it('should match snapshot', () => {
    const { container } = render(
      <AppLike>
        <SavingIndicator collabProvider={collabProvider} />
      </AppLike>
    )
    expect(container).toMatchSnapshot()
  })
}

describe('SavingIndicator', () => {
  afterEach(() => {
    jest.resetAllMocks()
  })
  beforeEach(() => {
    useBreakpoints.mockImplementation(() => ({
      isMobile: false
    }))

    // createRange is used by the MUI Tooltip for its placement
    global.document.createRange.mockImplementation(() => ({
      setStart: () => {},
      setEnd: () => {},
      commonAncestorContainer: {
        nodeName: 'BODY',
        ownerDocument: document
      }
    }))
  })

  describe('when dirtySince is truthy', () => {
    describe('when dirtySince is a few minutes ago', () => {
      beforeEach(() => {
        getDirtySince.mockImplementation(() => new Date(Date.now() - 15 * min))
      })
      itMatchSnapshot(collabProvider)
    })
    describe('when dirtySince is a few seconds ago', () => {
      beforeEach(() => {
        getDirtySince.mockImplementation(() => new Date(Date.now() - 2 * sec))
      })
      itMatchSnapshot(collabProvider)
    })
  })

  describe('when dirtySince is falsy', () => {
    const steps = {
      'few seconds': 2 * sec,
      '45 sec': 45 * sec,
      '1 min': 1 * min,
      '5 min': 5 * min,
      '15 min': 15 * min,
      '28 min': 28 * min,
      '50 min': 50 * min,
      '1 hour': 1 * hour,
      '3 hour': 3 * hour,
      '1 day': 1 * day,
      '10 day': 10 * day
    }
    Object.keys(steps).forEach(key => {
      describe(`when lastSave is ${key} ago`, () => {
        beforeEach(() => {
          getLastSaveOrSync.mockImplementation(
            () => new Date(Date.now() - steps[key] - 500)
          )
        })
        itMatchSnapshot(collabProvider)
      })
    })
  })

  describe('when offline', () => {
    beforeEach(() => {
      jest.spyOn(global.navigator, 'onLine', 'get').mockReturnValue(false)
    })
    // tooltip should be here
    itMatchSnapshot(collabProvider)
  })

  describe('when online', () => {
    beforeEach(() => {
      jest.spyOn(global.navigator, 'onLine', 'get').mockReturnValue(true)
    })
    // tooltip shouldn't be here
    itMatchSnapshot(collabProvider)
  })
})
