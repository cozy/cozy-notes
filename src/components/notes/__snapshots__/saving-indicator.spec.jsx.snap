// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SavingIndicator when dirtySince is falsy when lastSave is 1 day ago should match snapshot 1`] = `
<div>
  <div
    class="u-text saving-indicator"
  >
    Notes.SavingIndicator.saved.default
  </div>
</div>
`;

exports[`SavingIndicator when dirtySince is falsy when lastSave is 1 hour ago should match snapshot 1`] = `
<div>
  <div
    class="u-text saving-indicator"
  >
    Notes.SavingIndicator.saved.default
  </div>
</div>
`;

exports[`SavingIndicator when dirtySince is falsy when lastSave is 1 min ago should match snapshot 1`] = `
<div>
  <div
    class="u-text saving-indicator"
  >
    Notes.SavingIndicator.saved.default
  </div>
</div>
`;

exports[`SavingIndicator when dirtySince is falsy when lastSave is 3 hour ago should match snapshot 1`] = `
<div>
  <div
    class="u-text saving-indicator"
  >
    Notes.SavingIndicator.saved.default
  </div>
</div>
`;

exports[`SavingIndicator when dirtySince is falsy when lastSave is 5 min ago should match snapshot 1`] = `
<div>
  <div
    class="u-text saving-indicator"
  >
    Notes.SavingIndicator.saved.default
  </div>
</div>
`;

exports[`SavingIndicator when dirtySince is falsy when lastSave is 10 day ago should match snapshot 1`] = `
<div>
  <div
    class="u-text saving-indicator"
  >
    Notes.SavingIndicator.saved.default
  </div>
</div>
`;

exports[`SavingIndicator when dirtySince is falsy when lastSave is 15 min ago should match snapshot 1`] = `
<div>
  <div
    class="u-text saving-indicator"
  >
    Notes.SavingIndicator.saved.default
  </div>
</div>
`;

exports[`SavingIndicator when dirtySince is falsy when lastSave is 28 min ago should match snapshot 1`] = `
<div>
  <div
    class="u-text saving-indicator"
  >
    Notes.SavingIndicator.saved.default
  </div>
</div>
`;

exports[`SavingIndicator when dirtySince is falsy when lastSave is 45 sec ago should match snapshot 1`] = `
<div>
  <div
    class="u-text saving-indicator"
  >
    Notes.SavingIndicator.saved.default
  </div>
</div>
`;

exports[`SavingIndicator when dirtySince is falsy when lastSave is 50 min ago should match snapshot 1`] = `
<div>
  <div
    class="u-text saving-indicator"
  >
    Notes.SavingIndicator.saved.default
  </div>
</div>
`;

exports[`SavingIndicator when dirtySince is falsy when lastSave is few seconds ago should match snapshot 1`] = `
<div>
  <div
    class="u-text saving-indicator"
  >
    Notes.SavingIndicator.saved.default
  </div>
</div>
`;

exports[`SavingIndicator when dirtySince is truthy when dirtySince is a few minutes ago should match snapshot 1`] = `
<div>
  <div
    aria-describedby="mui-00000"
    class="u-text saving-indicator"
  >
    Notes.SavingIndicator.saving_impossible.just_now
  </div>
</div>
`;

exports[`SavingIndicator when dirtySince is truthy when dirtySince is a few seconds ago should match snapshot 1`] = `
<div>
  <div
    class="u-text saving-indicator"
  >
    Notes.SavingIndicator.saving
  </div>
</div>
`;

exports[`SavingIndicator when offline should match snapshot 1`] = `
<div>
  <div
    aria-describedby="mui-00000"
    class="u-text saving-indicator"
  >
    Notes.SavingIndicator.out_of_sync.just_now
  </div>
</div>
`;

exports[`SavingIndicator when online should match snapshot 1`] = `
<div>
  <div
    class="u-text saving-indicator"
  >
    Notes.SavingIndicator.saved.default
  </div>
</div>
`;
